src/mame/jaleco/jaleco_vj_qtaro.cpp

```diff
@@ -0,0 +1,653 @@
+}
+
+void jaleco_vj_qtaro_device::device_reset()
+{
+    m_int = 0;
+    m_mix_level = 0;
+
+    reset_video();
+}
+
+void jaleco_vj_qtaro_device::reset_video()
+{
+    m_frame_width = 352;
+    m_frame_height = 240;
+
+    if (m_raw_video != nullptr)
+        free(m_raw_video);
+
+    if (m_plm_video != nullptr)
+        plm_video_destroy(m_plm_video);
+
+    if (m_plm_buffer != nullptr)
+        plm_buffer_destroy(m_plm_buffer);
+
+    m_plm_buffer = plm_buffer_create_with_capacity(m_buffer_size);
+    m_plm_video = plm_video_create_with_buffer(m_plm_buffer, false);
+    m_raw_video = (uint8_t*)calloc(m_frame_width * m_frame_height * 4, sizeof(uint8_t));
+}
+
+void jaleco_vj_qtaro_device::mix_w(offs_t offset, uint16_t data)
+{
+    if (data != m_mix_level) {
+        LOGMASKED(LOG_VIDEO, "[%s] mix_w %04x\n", tag(), data);
+    }
+
+    m_mix_level = data;
+}
+
+uint8_t jaleco_vj_qtaro_device::reg_r(offs_t offset)
+{
+    return m_int;
+}
+
+void jaleco_vj_qtaro_device::reg_w(offs_t offset, uint8_t data)
+{
+    // Bit 7 is set when starting DMA for an entirely new video, then unset when video is ended
+    if (BIT(data, 7) && !BIT(m_int, 7))  {
+        LOGMASKED(LOG_VIDEO, "[%s] DMA transfer thread started %02x\n", tag(), data);
+        reset_video();
+    } else if (!BIT(data, 7) && BIT(m_int, 7))  {
+        LOGMASKED(LOG_VIDEO, "[%s] DMA transfer thread ended %02x\n", tag(), data);
+    }
+
+    m_int = data;
+}
+
+uint8_t jaleco_vj_qtaro_device::reg2_r(offs_t offset)
+{
+    // WriteStream cleanup function will loop until this returns 0.
+    // Probably relates to DMA or video playback state.
+    return 0;
+}
+
+uint32_t jaleco_vj_qtaro_device::reg3_r(offs_t offset)
+{
+    // 0x20 is some kind of default state. Relates to DMA or video playback.
+    // If this value is 0x40 then the code sets it back to 0x20 during the WriteStream cleanup function.
+    return 0x20;
+}
+
+void jaleco_vj_qtaro_device::reg3_w(offs_t offset, uint32_t data)
+{
+}
+
+void jaleco_vj_qtaro_device::write(uint8_t *data, uint32_t len)
+{
+    // LOGMASKED(LOG_VERBOSE_VIDEO_DATA, "[%s] video data write %02x\n", tag(), data);
+    plm_buffer_write(m_plm_buffer, data, len);
+}
+
+void jaleco_vj_qtaro_device::update_frame()
+{
+    plm_frame_t *frame = plm_video_decode(m_plm_video);
+
+    if (frame == nullptr)
+        return;
+
+    LOGMASKED(LOG_VIDEO,
+        "[%s] video frame generated %lf %d %08lx %08lx %08lx\n",
+        tag(),
+        plm_video_get_time(m_plm_video),
+        frame != nullptr,
+        plm_buffer_get_size(m_plm_buffer),
+        plm_buffer_get_remaining(m_plm_buffer),
+        get_remaining_memory_size()
+    );
+
+    if (m_frame_width != frame->width || m_frame_height != frame->height) {
+        if (m_raw_video != nullptr)
+            free(m_raw_video);
+
+        m_frame_width = frame->width;
+        m_frame_height = frame->height;
+        m_raw_video = (uint8_t*)calloc(m_frame_width * m_frame_height * 4, sizeof(uint8_t));
```

Comment:
I don't particularly like this approach of using `free` / `calloc`, isn't this `m_raw_video` practically a glorious `bitmap_rgb32` buffer?

Reply:
It could be a `bitmap_rgb32` but I didn't want to recreate a new `bitmap_rgb32` every frame read and I wasn't sure of a better way to do it. Is there a safe way to change the entire underlying buffer's data for a `bitmap_rgb32`? Would just getting the start of the array using something like `raw_pixptr(0, 0)` and then passing that into `plm_frame_to_bgra` be ok?

Reply:
You can clear a bitmap buffer by `bitmap_name.fill(0, cliprect) `

Reply:
It wouldn't need to be cleared. The entire contents of the bitmap need to be overwritten with the contents of the new MPEG video frame data.

Reply:
Yep, that looks better 👍🏽